testcases/elichika_tests/syntax/With.py

```diff
@@ -0,0 +1,110 @@
+# coding: utf-8
+
+import chainer
+from chainer_compiler.elichika.parser import flags
+
+class X(object):
+    def __init__(self, a):
+        self.a = a
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        return False
+
+class Y(X):
+    def __enter__(self):
+        self.a = self.a + 1
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.a = self.a + 2
+        return False
+
+class Z(X):
+    def __enter__(self):
+        return self.a
+
+# ======================================
+
+class A(chainer.Chain):
+    def forward(self):
+        with X(3) as obj:
+            return obj.a
+
+class B(chainer.Chain):
+    def forward(self):
+        ret = 0
+        with X(3) as obj1, X(4) as obj2:
+            ret += obj1.a
+        ret += obj2.a
+        return ret
+
+class C(chainer.Chain):
+    def forward(self):
+        ret = 0
+        obj2 = Y(4)
+        with Y(3) as obj1:
+            ret += obj1.a
+        ret +=  obj2.a
+        return ret
+
+class D(chainer.Chain):
+    def forward(self):
+        with Z(3) as obj:
+            return obj
+
+def func_elichika_dislike():
+    yield 42
+
+class IgnoreBranch(chainer.Chain):
+    def forward(self):
+        ret = 0
+        with flags.ignore_branch():
+            func_elichika_dislike()
+        ret += 1
+        return ret
+
+class ForUnroll(chainer.Chain):
+    def __init__(self):
+        super(ForUnroll, self).__init__()
+        self.x1 = 1
+        self.x2 = 2
+        self.x3 = 3
+        self.dict = {'x1': 1, 'x2': 2, 'x3': 3}
+
+    def forward(self):
+        ret = 0
+
+        with flags.for_unroll():
```

Comment:
Can we extend `for_unroll` to take an optional boolean argument which makes it possible for users to go back to normal mode? A usecase scenario in my mind:

```
with flags.for_unroll():
  hidden = inputs  # A list with `seq_length` elements.
  for layer_id in range(num_layers):  # num_layers is a constant
    next_hidden = []
    with flags.for_unroll(unroll=False):
      state = rnn_initial_state()
      for time in range(seq_length):  # seq_length is not a constant
         next_h, state = self['rnn_%d' % layer_id](hidden[time], state)
         next_hidden.append(next_h)
    hidden = next_hidden
```

Reply:
I added `unroll=False`. The above example however won't work without inside loop being unrolled as `'rnn_%d' % layer_id` requires constant propagation which would only work with unrolled for loop currently.

Reply:
Is this easy to do constant propagation in outside `for_unroll`?